File: synacor.py
#!/usr/bin/env python3

import array
import sys
import argparse

# 0000 : intro message
# 015b : self-test
# 0375 : call routine #1 (once)
# 05b2 : routine 05b2 - reads word at [8000], calls [8001]
# 05fb : routine 05fb - outputs text
# 06bb : routine 06bb (once)
# 06cb :   call routine 084d
# 06d0 :   call routine 084d
# 0706 : in op for prompt
# 084d : routine 084d (48187 times) (decryption?)
# 154b : check for hacked H register
# 154e : text announcing that confirmation routine is starting
# 155e :   call routine 05b2
# 1566 :   5 NOPs
# 156b :   2 SETs to 8000 and 8001 -> set to 1,4
# 1571 :   call to mega-recursive confirmation routine
# 1573 :   check that 8000==6
# 158a :   call routine 05b2
# 158c :   do something with 8007!
# 15a0 :   call routine 0731
# 15e3 :   jump to function epilogue
# 1652 :   three POP's and a RET
# 178b : mega-recursive confirmation routine (it's not Ackermann is it?)
#           (that's the most evilly recursive function I know, anyhow)
#           (i'll just bet that it is! >_<)
#           A(1,4)==6
#           A(4,1)==65533=0xfffd

def opcode(c):
    def opcode(f):
        f.opcode = c
        return f
    return opcode

class Synacor:
    def __init__(self, filename):
        self.mem = self.load(filename)
        self.reg = {x: 0 for x in range(0x8000, 0x8008)}
        self.ip = 0
        self.stack = []
        self.ops = {f.__func__.opcode: (f, f.__func__.__code__.co_argcount-1)
                    for f in [getattr(self, k) for k in dir(self)]
                    if hasattr(f, '__func__') and hasattr(f.__func__, 'opcode')}
        self.debug_level = 0

    @staticmethod
    def load(filename):
        memory = array.array('H')
        with open(filename, 'rb') as f:
            memory.frombytes(f.read())
        memory.extend([0] * (0xffff-len(memory)))
        return memory

    def run(self, break_at=None):
        while 0 <= self.ip <= 0xffff:
            if break_at is not None and self.ip == break_at:
                break
            opcode = self.mem[self.ip]
            op, c = self.ops[opcode]
            args = [self.mem[offset] for offset in range(self.ip+1, self.ip+c+1)]
            next_ip = op(*args)
            if self.debug_level > 0:
                print('{0:04x}  {1:<4s}  {2:<14s}  {3}  {4}'.format(
                    self.ip,
                    op.__name__[3:],
                    ' '.join('{:04x}'.format(a) for a in args),
                    ' '.join('{}={:04x}'.format(k, self.reg[v]) for (k, v) in zip('abcdefgh', range(0x8000, 0x8008))),
                    '' if op != self.op_out else repr(chr(self[self.mem[self.ip+1]]))
                ), file=sys.stderr, flush=False)
            self.ip = next_ip if next_ip is not None else self.ip+c+1

    def dump_memory(self):
        for i in range(0, len(M), 8):
            print('{:08x} {:04x} {:04x} {:04x} {:04x} {:04x} {:04x} {:04x} {:04x}'.format(i, *[M[i+x] for x in range(8)]))

    def disassemble(self):
        ip = 0
        while ip < len(self.mem):
            try:
                op, c = self.ops[self.mem[ip]]
                args = ['{0:04x}'.format(self.mem[offset]) for offset in range(ip+1, ip+c+1)]
                print('{0:04x}  {1:<4s}  {2:<14s}  {3}'.format(
                    ip,
                    op.__name__[3:],
                    ' '.join(args),
                    '' if op != self.op_out else repr(chr(self[self.mem[ip+1]]))))
                ip += 1+c
            except KeyError:
                print('{0:04x}  {1:<4s}  {2:04x}  {3}'.format(
                    ip,
                    'dw',
                    self.mem[ip],
                    repr(chr(self.mem[ip])) if 32 <= self.mem[ip] < 128 else ''))
                ip += 1

    def poke(self, offset, *values):
        for (i, v) in enumerate(values):
            self.mem[offset+i] = v

    def __getitem__(self, k):
        return self.reg.get(k, k)

    def __setitem__(self, k, v):
        if not (0x8000 <= k < 0x8008):
            raise ValueError('invalid register {:04x}'.format(k))
        self.reg[k] = v

    @opcode(0x00)
    def op_halt(self):
        return -1

    @opcode(0x01)
    def op_set(self, a, b):
        self[a] = self[b]

    @opcode(0x02)
    def op_push(self, a):
        self.stack.append(self[a])

    @opcode(0x03)
    def op_pop(self, a):
        if not self.stack:
            raise RuntimeError('stack is empty')
        self[a] = self.stack.pop()

    @opcode(0x04)
    def op_eq(self, a, b, c):
        self[a] = 1 if self[b] == self[c] else 0

    @opcode(0x05)
    def op_gt(self, a, b, c):
        self[a] = 1 if self[b] > self[c] else 0

    @opcode(0x06)
    def op_jmp(self, a):
        return self[a]

    @opcode(0x07)
    def op_jt(self, a, b):
        return self[b] if self[a] != 0 else None

    @opcode(0x08)
    def op_jf(self, a, b):
        return self[b] if self[a] == 0 else None

    @opcode(0x09)
    def op_add(self, a, b, c):
        self[a] = (self[b] + self[c]) % 0x8000

    @opcode(0x0a)
    def op_mult(self, a, b, c):
        if b == 6 and c == 9 and 0: # set to 1 to trigger 'not hitchhiking' self-test error :-)
            self[a] = 42
            return
        self[a] = (self[b] * self[c]) % 0x8000

    @opcode(0x0b)
    def op_mod(self, a, b, c):
        self[a] = self[b] % self[c]

    @opcode(0x0c)
    def op_and(self, a, b, c):
        self[a] = self[b] & self[c]

    @opcode(0x0d)
    def op_or(self, a, b, c):
        self[a] = self[b] | self[c]

    @opcode(0x0e)
    def op_not(self, a, b):
        self[a] = self[b] ^ 0x7fff

    @opcode(0x0f)
    def op_rmem(self, a, b):
        self[a] = self.mem[self[b]]

    @opcode(0x10)
    def op_wmem(self, a, b):
        self.mem[self[a]] = self[b]

    @opcode(0x11)
    def op_call(self, a):
        self.stack.append(self.ip+2)
        return self[a]

    @opcode(0x12)
    def op_ret(self):
        if not self.stack:
            raise RuntimeError('stack is empty')
        return self.stack.pop()

    @opcode(0x13)
    def op_out(self, a):
        sys.stdout.write(chr(self[a]))
        sys.stdout.flush()

    @opcode(0x14)
    def op_in(self, a):
        c = sys.stdin.read(1)
        if not c:
            return -1
        self[a] = ord(c)

    @opcode(0x15)
    def op_noop(self):
        pass

#---

def test_load_file_1(): assert list(load('challenge.bin')[:3]) == [21,21,19]
def test_load_file_2(): assert len(load('challenge.bin')) == 65535

#---

def parse_args(args):
    parser = argparse.ArgumentParser()
    parser.add_argument('filename', help='the .bin file to load')
    parser.add_argument('--debug', '-d', action='count', default=0, help='debug level')
    parser.add_argument('--breakpoint', action='store', default=None, help='break at hexadecimal address')
    parser.add_argument('--poke', action='store', nargs='+', help='hexadecimal numbers, first is address, rest is machine code')
    group = parser.add_mutually_exclusive_group()

    group.add_argument('--dump', action='store_true', help='output memory dump')
    group.add_argument('--disasm', action='store_true', help='output disassembly')
    return parser.parse_args(args)

def main(args):
    args = parse_args(args)
    #print(args)
    synacor = Synacor(args.filename)
    synacor.debug_level = args.debug
    if args.poke:
        for values in args.poke:
            synacor.poke(*[int(x, 16) for x in values.split()])
    if args.breakpoint is not None:
        synacor.run(break_at=int(args.breakpoint, 16))
    if args.dump:
        synacor.dump_memory()
    elif args.disasm:
        synacor.disassemble()
    else:
        synacor.run()

if __name__ == '__main__':
    main(sys.argv[1:])

File: synacor.c

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>

struct vm_t {
    uint16_t *code;
    uint16_t *stack;
    int code_size;
    int stack_size;
};

#define A *(ip+1)
#define B *(ip+2)
#define C *(ip+3)

#define s(v) reg[v^0x8000]
#define r(v) (v&0x8000 ? s(v) : v)

#define CHK_SP_MIN if (sp == vm->stack) { fprintf(stderr, "error: stack is empty at %04x\n", (uint16_t)(ip-vm->code)); return 1; }
#define CHK_SP_MAX if (sp == sp_max)    { fprintf(stderr, "error: stack is full at %04x\n",  (uint16_t)(ip-vm->code)); return 1; }

int execute(struct vm_t *vm, int debug) {
    const char *op_name[22] = { "halt", "set", "push", "pop", "eq", "gt", "jmp", "jt",
                                "jf", "add", "mult", "mod", "and", "or", "not", "rmem",
                                "wmem", "call", "ret", "out", "in", "noop" };
    const int arg_count[22] = { 0, 2, 1, 1, 3, 3, 1, 2,
                                2, 3, 3, 3, 3, 3, 2, 2,
                                2, 1, 0, 1, 1, 0 };

    uint16_t reg[8] = {0, 0, 0, 0, 0, 0, 0, 0};

    uint16_t *ip = vm->code;
    uint16_t *ip_max = vm->code + vm->code_size - 1;

    uint16_t *sp = vm->stack;
    uint16_t *sp_max = vm->stack + vm->stack_size - 1;

    while (vm->code <= ip && ip < ip_max) {
        if (debug) {
            switch (arg_count[*ip]) {
                case 0: fprintf(stderr, "%04x  %-4s                \n", (uint16_t)(ip-vm->code), op_name[*ip]);          break;
                case 1: fprintf(stderr, "%04x  %-4s  %04x          \n", (uint16_t)(ip-vm->code), op_name[*ip], A);       break;
                case 2: fprintf(stderr, "%04x  %-4s  %04x %04x     \n", (uint16_t)(ip-vm->code), op_name[*ip], A, B);    break;
                case 3: fprintf(stderr, "%04x  %-4s  %04x %04x %04x\n", (uint16_t)(ip-vm->code), op_name[*ip], A, B, C); break;
            }
        }

        switch (*ip) {
            case 0x00: /* halt */       return 0;
            case 0x01: /* set a b */    s(A) = r(B);                            ip += 3; break;
            case 0x02: /* push a */     CHK_SP_MAX; *sp++ = r(A);               ip += 2; break;
            case 0x03: /* pop a */      CHK_SP_MIN; s(A) = *--sp;               ip += 2; break;
            case 0x04: /* eq a b c */   s(A) = r(B) == r(C) ? 1 : 0;            ip += 4; break;
            case 0x05: /* gt a b c */   s(A) = r(B) >  r(C) ? 1 : 0;            ip += 4; break;
            case 0x06: /* jmp a */      ip = vm->code + r(A);                            break;
            case 0x07: /* jt a b */     ip =  r(A) ? vm->code + r(B) : ip+3;             break;
            case 0x08: /* jf a b */     ip = !r(A) ? vm->code + r(B) : ip+3;             break;
            case 0x09: /* add a b c */  s(A) = (r(B) + r(C)) & 0x7fff;          ip += 4; break; 
            case 0x0a: /* mult a b c */ s(A) = (r(B) * r(C)) & 0x7fff;          ip += 4; break; 
            case 0x0b: /* mod a b c */  s(A) = r(B) % r(C);                     ip += 4; break; 
            case 0x0c: /* and a b c */  s(A) = r(B) & r(C);                     ip += 4; break; 
            case 0x0d: /* or a b c */   s(A) = r(B) | r(C);                     ip += 4; break; 
            case 0x0e: /* not a b */    s(A) = r(B) ^ 0x7fff;                   ip += 3; break; 
            case 0x10: /* wmem a b */   *(vm->code + r(A)) = r(B);              ip += 3; break; 
            case 0x0f: /* rmem a b */   s(A) = *(vm->code + r(B));              ip += 3; break; 
            case 0x11: /* call a */     CHK_SP_MAX; *sp++ = ip-vm->code+2;      ip = vm->code + r(A); break;
            case 0x12: /* ret */        CHK_SP_MIN; ip = vm->code + *--sp;               break;
            case 0x13: /* out a */      printf("%c", r(A));                     ip += 2; break;
            case 0x14: /* in a */     { char ch;
                                        int count = fread(&ch, 1, sizeof(ch), stdin);
                                        if (count == 0) {
                                            return 0;
                                        }
                                        s(A) = (uint16_t)ch;                    ip += 2; break; }
            case 0x15: /* noop */                                               ip += 1; break;
            default:                    fprintf(stderr, "error: unrecognized instruction %#04x!\n", *ip);
                                        return 1;
        }
    }

    return 0;
}

#undef A
#undef B
#undef C

#undef s
#undef r

int poke_code(struct vm_t *vm, char *pokes[], int poke_max) {
    for (int i = 0; i < poke_max && pokes[i]; i++) {
        uint16_t *dest = NULL;
        char *next = NULL;
        for (char *value = strtok_r(pokes[i], " ", &next); value != NULL; value = strtok_r(NULL, " ", &next)) {
            char *end = NULL;
            uint16_t word = strtoul(value, &end, 16);
            if (end == value || *end != '\0') {
                fprintf(stderr, "error: cannot parse value '%s'\n", value);
                return 1;
            }
            if (dest == NULL) {
                dest = vm->code + word;
            }
            else {
                *dest++ = word;
            }
        }
    }
    return 0;
}

int load_program(struct vm_t *vm, char *filename) {
    FILE *fp = fopen(filename, "rb");

    if (fp == NULL) {
        fprintf(stderr, "error: failed to open %s\n", filename);
        return 1;
    }

    uint16_t *offset = vm->code;
    int count;
    do {
        count = fread(offset, 1, vm->code_size-(offset-vm->code), fp);
        offset += count;
    } while (count && (offset-vm->code) < vm->code_size);

    return 0;
}

void free_vm(struct vm_t *vm) {
    free(vm->code);
    free(vm->stack);
    memset(vm, 0, sizeof(struct vm_t));
}

int init_vm(struct vm_t *vm) {
    vm->code_size = sizeof(uint16_t) * 0x8000;
    vm->code = malloc(vm->code_size);
    if (!vm->code) {
        goto error;
    }
    memset(vm->code, 0, vm->code_size);

    vm->stack_size = sizeof(uint16_t) * 0x80000;
    vm->stack = malloc(vm->stack_size);
    if (!vm->stack) {
        goto error;
    }
    memset(vm->stack, 0, vm->stack_size);

    return 0;

error:
    free_vm(vm);
    return 1;
}

int parse_options(int argc, char *argv[], char **filename, char *pokes[], int max_pokes, int *debug) {
    int opt;
    int poke_index = 0;
    while ((opt = getopt(argc, argv, "-p:d")) > -1) {
        switch (opt) {
            case 0x01:
                if (*filename != NULL) {
                    fprintf(stderr, "error: more than one filename specified\n");
                    return 1;
                }
                *filename = optarg;
                continue;
            case 'd':
                *debug = 1;
                continue;
            case 'p':
                if (poke_index == max_pokes - 1) {
                    fprintf(stderr, "error: more than %d pokes\n", max_pokes);
                    return 1;
                }
                pokes[poke_index++] = optarg;
                continue;
            case '?':
                return 1;
        }
    }

    if (*filename == NULL) {
        fprintf(stderr, "error: missing filename\n");
        return 1;
    }

    return 0;
}

int main(int argc, char *argv[]) {
    char *filename = NULL;
    char *pokes[16] = {0};
    int debug = 0;

    // parse options
    if (parse_options(argc, argv, &filename, pokes, sizeof(pokes), &debug)) {
        return 1;
    }

    // initialize vm memory
    struct vm_t vm = {0};
    if (init_vm(&vm)) {
        return 2;
    }

    // load program from file
    if (load_program(&vm, filename)) {
        goto error;
    }

    // poke memory if so requested
    if (poke_code(&vm, pokes, sizeof(pokes))) {
        goto error;
    }

    // run the code
    if (execute(&vm, debug)) {
        goto error;
    }

    // clean exit
    free_vm(&vm);
    return 0;

error:
    // error exit
    free_vm(&vm);
    return 1;
}
